src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -2138,15 +2159,58 @@ public void NotifyListenersAbout(long index, Exception e)
                     Index = index,
                     Exception = ExceptionDispatchInfo.Capture(e)
                 });
+
                 if (Interlocked.Increment(ref _numberOfErrors) > 25)
                 {
                     _errors.TryDequeue(out _);
                     Interlocked.Decrement(ref _numberOfErrors);
                 }
             }
-            ClusterStateMachine.InterlockedExchangeMax(ref LastModifiedIndex, index);
+
+            InterlockedExchangeMax(ref LastModifiedIndex, index);
             _notifiedListeners.SetAndResetAtomically();
         }
+
+        public void SetTaskCompleted(long index, Exception e)
+        {
+            if (_tasksDictionary.TryGetValue(index, out var tcs))
+            {
+                try
+                {
+                    // set the task as finished
+                    if (e == null)
+                        tcs.SetResult(null);
+                    else
+                        tcs.SetException(e);
+                }
+                catch (Exception ex)
+                {
+                    if (Log.IsInfoEnabled)
+                        Log.Info($"Failed to set result of task with index {index}", ex);
+                }
+            }
+
+            _tasksDictionary.TryRemove(index, out _);
```

Comment:
Where are we putting the error in the queue?

Reply:
For that matter, shouldn't we just use the dictionary, instead of doing linear search?
We can keep the tasks there until they are either:
* done (no error)
* done with error and observed
* done with error, unobserved and too much time have passed

Reply:
There are a couple of scenarios:
- We are waiting on the task (no error)
- We are waiting on the task (with error)
- We start waiting after the task was already completed (and removed from the dictionary).
In that case we look in the last errors queue (this is the only case that we do that):
https://github.com/ravendb/ravendb/blob/5e5be6d9c3da176b04242dfb760bdca526440f71/src/Raven.Server/ServerWide/ClusterStateMachine.cs#L2076